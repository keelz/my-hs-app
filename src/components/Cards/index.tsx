import * as React from 'react';
import Card from '../Card';
import './Cards.css';
import {
  Card as CardType,
  CardResolution,
  CardExt,
  CardLocale
} from '../../redux/reducers/Cards';

interface CardsProps {}

export interface CardsStateProps {
  activeCardClassName: string;
  cards: CardType[];
}

type Props = CardsProps & CardsStateProps;

const Cards: React.SFC<Props> = props =>
  <div className="Cards">
    <div className="Cards-title">
      {props.activeCardClassName}: {props.cards.length}
    </div>
    <div className="Cards-collection">
      { props.cards.map(card =>
        <Card
          key={card.id}
          ext={CardExt.PNG}
          id={card.id}
          locale={CardLocale.EN}
          resolution={CardResolution.SMALL} />)
      }
    </div>
  </div>;

export default Cards;
