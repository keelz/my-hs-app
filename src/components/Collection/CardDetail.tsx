import * as React from 'react';
import * as classnames from 'classnames';
import HSJSON from '../../common/constants/hsJson';
import { IComponentProps } from '../../common/models/App.model';
import { composeClassname } from '../../common/utils';
import { composeAssetSource } from '../../common/services/hsJson.service';
import {
  CardExt,
  CardLocale,
  CardResolution,
  CardSet,
  ICard,
  CardType,
  Race,
  Rarity,
} from '../../common/models/Cards.model';

interface ICardDetailProps extends IComponentProps {
  card?: ICard;
}

const composeCardSetName = (cardSet: CardSet): String =>
  HSJSON.RESPONSE_ENUM.CARD_SET[cardSet];

const composeCardTypeName = (cardType: CardType): String =>
  HSJSON.RESPONSE_ENUM.CARD_TYPE[cardType];

const composeCardRaceName = (race?: Race): String =>
  !!race ? HSJSON.RESPONSE_ENUM.RACE[race] : '';

const composeCardRarityName = (rarity?: Rarity): String =>
  !!rarity ? HSJSON.RESPONSE_ENUM.RARITY[rarity] : '';

const CardDetail: React.SFC<ICardDetailProps> = props =>
  <div className={composeClassname([
    'App-card-detail',
    'flex',
    'flex-row'])(props.className)}>
    <div style={{ width: 340, backgroundColor: 'transparent' }}>
      <img style={{ marginTop: -60 }} src={composeAssetSource(
        props.card!.id,
        CardLocale.EN,
        CardResolution.LARGE,
        CardExt.PNG
      )} />
    </div>
    <div className={classnames([
      'bg-grey-darkest',
      'flex-grow',
      'rounded-lg',
      'shadow-outline',
      'text-white',
    ])} style={{ padding: 10, width: 300 }}>
      <h2>{props.card!.name}</h2>
      <p className="mt-2 mb-2">{props.card!.flavor}</p>
      <p className="mt-2 mb-2">Set: {composeCardSetName(props.card!.set)}</p>
      <p className="mt-2 mb-2">Type: {composeCardTypeName(props.card!.type)}</p>
      { props.card && props.card.race &&
        <p className="mt-2 mb-2">Race: {composeCardRaceName(props.card!.race)}</p>
      }
      { props.card && props.card.rarity &&
        <p className="mt-2 mb-2">Rarity: {composeCardRarityName(props.card.rarity)}</p>
      }
    </div>
  </div>;

export default CardDetail;
