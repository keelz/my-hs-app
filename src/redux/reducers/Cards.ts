import { Dispatch } from 'react';
import { IActionType } from '../Types';
import { ICard } from '../../common/models/Cards.model';

/** MODEL */
export interface ICardsState {
  data: ICard[];
}

/** STATE FACTORY */
export const defaultState = (): ICardsState => ({
  data: Array(0),
});

/** ACTIONS */
export const CARDS_SET_CARDS = 'CARDS_SET_CARDS';

/** ACTION CREATORS */
export const cardsSetCardsAction = (data: ICard[]) =>
  (dispatch: Dispatch<IActionType<any>>) => {
    dispatch({
      type: CARDS_SET_CARDS,
      payload: { data },
    });
  };

/** REDUCER */
export const reducer = (
  state: any = defaultState(),
  action: IActionType<ICardsState>
): ICardsState => {
  const { type } = action;
  switch (type) {
    case CARDS_SET_CARDS: return {
      ...state,
      data: action.payload!.data,
    };
    default: return state;
  }
};

export default reducer;
