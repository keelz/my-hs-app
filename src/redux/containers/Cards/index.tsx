import { connect } from 'react-redux';
import { createSelector } from 'reselect';
import Cards from '../../../components/Cards';
import { RootState } from '../..//Types';
import { Card } from '../../reducers/Cards';

const getCards = (state: RootState) => state.Cards.data;

const getActiveClassName = (state: RootState) => state.Cards.activeClassName;

const cardsSelector = createSelector(
  [getCards, getActiveClassName],
  (cards: Card[], activeClassName) =>
    cards
      .filter(c => c.cardClass === activeClassName)
      .sort((a, b) => a.name < b.name ? - 1 : 0)
);

const mapStateToProps = (state: RootState) => ({
  activeCardClassName: getActiveClassName(state),
  cards: cardsSelector(state),
});

const ConnectedCards = connect(mapStateToProps)(Cards);

export default ConnectedCards;
