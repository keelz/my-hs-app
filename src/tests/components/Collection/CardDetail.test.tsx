import * as React from 'react';
import * as enzyme from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import CardDetail from '../../../components/Collection/CardDetail';
import { CardClassName, CardSet, CardType, Race, Rarity } from '../../../common/models/Cards.model';

const defaultProps = Object.freeze({
  card: {
    artist: 'test-artist',
    cardClass: CardClassName.DRUID,
    collectible: true,
    cost: 7,
    dbfId: 123,
    flavor: 'test-card-flavor',
    id: 'test-card-id',
    name: 'test-card-name',
    race: Race.BLOODELF,
    rarity: Rarity.EPIC,
    set: CardSet.BOOMSDAY,
    text: 'test-card-text',
    type: CardType.MINION,
  },
});

describe('Collecttion/CardDetail', () => {
  describe('snapshots', () => {
    it('renders without crashing', () => {
      const tree = enzyme.shallow(<CardDetail {...defaultProps} />);
      expect(shallowToJson(tree)).toMatchSnapshot();
    });
  });
});
