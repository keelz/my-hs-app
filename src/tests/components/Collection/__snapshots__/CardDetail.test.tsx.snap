// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collecttion/CardDetail snapshots renders without crashing 1`] = `
<div
  className="App-card-detail flex flex-row"
>
  <div
    style={
      Object {
        "backgroundColor": "transparent",
        "width": 340,
      }
    }
  >
    <img
      src="https://art.hearthstonejson.com/v1/render/latest/enUS/512x/test-card-id.png"
      style={
        Object {
          "marginTop": -60,
        }
      }
    />
  </div>
  <div
    className="bg-grey-darkest flex-grow rounded-lg shadow-outline text-white"
    style={
      Object {
        "padding": 10,
        "width": 300,
      }
    }
  >
    <h2>
      test-card-name
    </h2>
    <p
      className="mt-2 mb-2"
    >
      test-card-flavor
    </p>
    <p
      className="mt-2 mb-2"
    >
      Set: 
      Booms Day
    </p>
    <p
      className="mt-2 mb-2"
    >
      Type: 
      MINION
    </p>
    <p
      className="mt-2 mb-2"
    >
      Race: 
      Blood Elf
    </p>
    <p
      className="mt-2 mb-2"
    >
      Rarity: 
      EPIC
    </p>
  </div>
</div>
`;
